subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/alignment/AlignmentIntegrationTest.groovy

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.resolve.alignment
+
+import org.gradle.integtests.fixtures.GradleMetadataResolveRunner
+import org.gradle.integtests.resolve.AbstractModuleDependencyResolveTest
+
+class AlignmentIntegrationTest extends AbstractModuleDependencyResolveTest {
+
+
+    def "should align leaves to the same version"() {
+        repository {
+            path 'xml -> core'
+            path 'json -> core'
+            path 'xml:1.1 -> core:1.1'
+            path 'json:1.1 -> core:1.1'
+        }
+
+        given:
+        buildFile << """
+            dependencies {
+                conf 'org:xml:1.0'
+                conf 'org:json:1.1'
+            }
+        """
+        and:
+        "a rule which infers module set from group and version"()
+
+        when:
+        repositoryInteractions {
+            'org:core:1.0' {
+                expectGetMetadata()
+            }
+            'org:xml:1.0' {
+                expectGetMetadata()
+            }
+            'org:core:1.1' {
+                expectResolve()
+            }
+            'org:json:1.1' {
+                expectResolve()
+            }
+            'org:xml:1.1' {
+                expectResolve()
+            }
+            'org:platform:1.0' {
+                expectGetMetadataMissing()
+                if (!GradleMetadataResolveRunner.experimentalResolveBehaviorEnabled) {
+                    expectHeadArtifactMissing()
+                }
+            }
+            'org:platform:1.1' {
+                expectGetMetadataMissing()
```

Comment:
I agree, and it's already visible when I play with the feature: I can see we're trying again the same versions build after build. I'm not quite sure how to solve this problem without compromising the feature itself.

Reply:
The idea of "retry without the cache" was to make a best attempt rather than failing the build. Since in this case the missing metadata will not fail the build, we could avoid the retry.

It might be tricky to wire this into here, though: https://github.com/gradle/gradle/blob/a275fe94c99e09b6360d8ebe83448223905cf638/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java#L148-L150

Reply:
I agree. In practice, we don't even need to _fetch_ the metadata. The only thing we need is to determine that the version exists. Fetching can be done in a subsequent step. So I wonder if we should provide a different "resolver" just for that purpose.